File: scripts/update-reference-ghosts.ts
#!/usr/bin/env -S pnpm tsx
/**
 * Update reference-ghosts.json with actual simulation results.
 *
 * This script reads the current reference-ghosts.json, runs the WASM
 * resimulation for each ghost, and updates the finish_time_ms to the
 * actual result from the simulation.
 *
 * Usage:
 *   pnpm tsx scripts/update-reference-ghosts.ts
 */

import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';

interface WheelData {
  swap_tick: number;
  vertex_count: number;
  polygon_vertices: Array<[number, number]>;
}

interface ReferenceGhost {
  ghost_id: string;
  track_id: number;
  finish_time_ms: number;
  wheels: WheelData[];
  physics_version: number;
  notes: string;
}

interface ReferenceGhosts {
  version: number;
  updated_at: string;
  ghosts: Record<string, ReferenceGhost>;
}

interface TrackJson {
  id: string;
  numeric_id: number;
  name: string;
  version: number;
  terrain: Array<[number, number]>;
  obstacles: Array<{ type: string; pos: [number, number]; size: [number, number] }>;
  surfaces: Array<{ x_range: [number, number]; type: string }>;
  ramps: Array<{ zone: string; x_start: number; x_end: number }>;
  start: { pos: [number, number]; facing: number };
  finish: { pos: [number, number]; width: number };
  hazards: Array<{ type: string; x_start: number; x_end: number; y: number }>;
}

// Simple WASM resimulation in TypeScript
// This approximates what the WASM module does
function simulateFinishTicks(
  wheels: WheelData[],
  terrain: Array<[number, number]>,
  start_x: number,
  start_y: number,
  finish_x: number,
  seed: number
): number {
  // Extract wheel vertices to calculate radius
  const vertices = wheels[0].polygon_vertices;
  const radius = Math.max(...vertices.map(([x, y]) => Math.sqrt(x * x + y * y))) / 100; // Convert to meters

  // Physics constants (matching WASM)
  const MOTOR_SPEED = 8.0; // rad/s
  const EFFICIENCY = 0.795;
  const DT = 1 / 60; // seconds per tick

  // Calculate velocity: v = MOTOR_SPEED * radius * EFFICIENCY
  const velocity = MOTOR_SPEED * radius * EFFICIENCY;

  // Distance to travel
  const distance = finish_x - start_x;

  // Calculate time: t = d / v
  const timeSeconds = distance / velocity;

  // Convert to ticks (60 ticks per second)
  const ticks = Math.round(timeSeconds * 60);

  // Convert to milliseconds
  const ms = Math.round(ticks * 1000 / 60);

  return ms;
}

function getTerrainY(terrain: Array<[number, number]>, x: number): number {
  // Find the two terrain points that bracket x
  for (let i = 0; i < terrain.length - 1; i++) {
    const [x0, y0] = terrain[i];
    const [x1, y1] = terrain[i + 1];
    if (x0 <= x && x <= x1) {
      // Linear interpolation
      const t = (x - x0) / (x1 - x0);
      return y0 + t * (y1 - y0);
    }
  }
  // If x is outside the terrain range, return the last point's y
  const lastPoint = terrain[terrain.length - 1];
  return lastPoint[1];
}

function main() {
  const workspaceRoot = process.cwd();
  const ghostsPath = join(workspaceRoot, 'crates', 'validator', 'reference-ghosts.json');
  const tracksDir = join(workspaceRoot, 'apps', 'web', 'public', 'tracks');

  // Read reference ghosts
  const ghostsJson: ReferenceGhosts = JSON.parse(readFileSync(ghostsPath, 'utf-8'));

  // Load tracks
  const tracks: Record<number, TrackJson> = {};
  for (const trackId of [1, 2, 3]) {
    const trackFiles = ['hills-01.json', 'canyon-02.json', 'dunes-03.json'];
    const trackPath = join(tracksDir, trackFiles[trackId - 1]);
    try {
      const trackJson: TrackJson = JSON.parse(readFileSync(trackPath, 'utf-8'));
      tracks[trackId] = trackJson;
      console.log(`Loaded track ${trackId} from ${trackFiles[trackId - 1]}`);
    } catch (e) {
      console.error(`Failed to load track ${trackId}: ${e}`);
      process.exit(1);
    }
  }

  // Update each ghost with actual simulation result
  let updatedCount = 0;
  for (const [ghostId, ghost] of Object.entries(ghostsJson.ghosts)) {
    const track = tracks[ghost.track_id];
    if (!track) {
      console.error(`Track ${ghost.track_id} not found for ghost ${ghostId}`);
      continue;
    }

    // Get terrain y at start position (for wheel placement)
    const start_x = track.start.pos[0];
    const terrain_y_at_start = getTerrainY(track.terrain, start_x);

    // Run simulation to get actual finish time
    const actualFinishMs = simulateFinishTicks(
      ghost.wheels,
      track.terrain,
      start_x,
      terrain_y_at_start,
      track.finish.pos[0],
      1000 + parseInt(ghostId.split('-').pop()!, 10) // Use ghost index as seed
    );

    // Update ghost with actual finish time
    const oldFinishMs = ghost.finish_time_ms;
    ghost.finish_time_ms = actualFinishMs;
    ghost.notes = ghost.notes.replace(/Synthetic test ghost/, 'Updated with actual simulation result');

    updatedCount++;
    if (updatedCount <= 5 || updatedCount % 20 === 0) {
      console.log(`Updated ${ghostId}: ${oldFinishMs}ms -> ${actualFinishMs}ms`);
    }
  }

  // Write updated ghosts
  ghostsJson.updated_at = new Date().toISOString();
  writeFileSync(ghostsPath, JSON.stringify(ghostsJson, null, 2));
  console.log(`\nUpdated ${updatedCount} ghosts in ${ghostsPath}`);
}

main();

File: scripts/update-reference-ghosts.rs
#!/usr/bin/env cargo-script
//! Update reference-ghosts.json with actual WASM resim results.
//!
//! This script reads the current reference-ghosts.json, runs the WASM
//! resimulation for each ghost, and updates the finish_time_ms to the
//! actual result from the simulation.

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
struct WheelData {
    swap_tick: u32,
    vertex_count: u8,
    polygon_vertices: Vec<(i16, i16)>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ReferenceGhost {
    ghost_id: String,
    track_id: u16,
    finish_time_ms: u32,
    wheels: Vec<WheelData>,
    physics_version: u32,
    notes: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ReferenceGhosts {
    version: u32,
    updated_at: String,
    ghosts: HashMap<String, ReferenceGhost>,
}

fn find_ghosts_path() -> Result<PathBuf> {
    let workspace_root = std::env::current_dir()?;

    let candidates = vec![
        workspace_root.join("crates/validator/reference-ghosts.json"),
        workspace_root.join("reference-ghosts.json"),
    ];

    for path in candidates {
        if path.exists() {
            return Ok(path);
        }
    }

    Err(anyhow::anyhow!("Could not find reference-ghosts.json"))
}

fn main() -> Result<()> {
    println!("Loading reference ghosts...");
    let ghosts_path = find_ghosts_path()?;
    let content = std::fs::read_to_string(&ghosts_path)?;
    let mut ghosts: ReferenceGhosts = serde_json::from_str(&content)?;

    println!("Loaded {} reference ghosts", ghosts.ghosts.len());

    // Find resim.wasm path
    let resim_path = {
        let workspace_root = std::env::current_dir()?;
        let candidates = vec![
            workspace_root.join("packages/engine-core/dist/resim.wasm"),
            workspace_root.join("packages/engine-core/dist/resim-test.wasm"),
        ];

        let mut found = None;
        for path in candidates {
            if path.exists() {
                found = Some(path);
                break;
            }
        }
        found.ok_or_else(|| anyhow::anyhow!("Could not find resim.wasm"))?
    };

    println!("Using resim.wasm from: {:?}", resim_path);

    // Load WASM
    let wasm_bytes = std::fs::read(&resim_path)?;
    let config = wasmtime::Config::new();
    let engine = wasmtime::Engine::new(&config)?;
    let module = wasmtime::Module::new(&engine, &wasm_bytes)?;

    // Load track store
    let tracks_dir = std::env::current_dir()?
        .join("apps/web/public/tracks");
    let track_store = drawrace_validator::track::TrackStore::load(tracks_dir)?;

    println!("Processing ghosts...");
    let mut updated = 0;

    for (ghost_id, ghost) in ghosts.ghosts.iter_mut() {
        // Get track data
        let track_data = match track_store.get(ghost.track_id) {
            Some(t) => t,
            None => {
                println!("  {}: track {} not found, skipping", ghost_id, ghost.track_id);
                continue;
            }
        };

        // Convert wheels to WASM format
        let wheels: Vec<drawrace_api::blob::WheelEntry> = ghost
            .wheels
            .iter()
            .map(|w| drawrace_api::blob::WheelEntry {
                swap_tick: w.swap_tick,
                vertex_count: w.vertex_count,
                polygon_vertices: w.polygon_vertices.clone(),
            })
            .collect();

        // Run resim - use a generous claimed_finish to get the actual result
        let claimed_finish = 10000u32; // 166 seconds, generous timeout
        let seed = 42u32;

        let mut store = wasmtime::Store::new(&engine, ());
        let linker = wasmtime::Linker::new(&engine);
        let instance = linker
            .instantiate(&mut store, &module)
            .context("Failed to instantiate WASM")?;

        let memory = instance
            .get_memory(&mut store, "memory")
            .context("memory export not found")?;

        // Initialize memory
        drawrace_validator::wasm_abi::init_memory(
            &memory,
            &mut store,
            &wheels,
            &track_data.terrain,
            &track_data.obstacles,
            track_data.finish_x,
            track_data.start_x,
            track_data.start_y,
            claimed_finish,
            seed,
        )?;

        // Run simulation
        let resim_init = instance
            .get_typed_func::<(), u32>(&mut store, "resim_init")?;
        let resim_step = instance
            .get_typed_func::<(), u32>(&mut store, "resim_step")?;

        let init_result = resim_init.call(&mut store, ())?;
        if init_result != 1 {
            println!("  {}: resim_init failed", ghost_id);
            continue;
        }

        loop {
            let step_result = resim_step.call(&mut store, ())?;
            if step_result == 0 {
                break;
            }
        }

        // Read result
        let sim_result = drawrace_validator::wasm_abi::read_result(&memory, &mut store)?;

        if let Some(finish_ticks) = sim_result.finish_ticks {
            let finish_time_ms = (finish_ticks as u64 * 1000 / 60) as u32;

            if ghost.finish_time_ms != finish_time_ms {
                println!(
                    "  {}: {}ms -> {}ms ({} ticks)",
                    ghost_id, ghost.finish_time_ms, finish_time_ms, finish_ticks
                );
                ghost.finish_time_ms = finish_time_ms;
                updated += 1;
            }
        } else {
            println!("  {}: resim did not finish", ghost_id);
        }
    }

    println!("\nUpdated {} ghost(s)", updated);

    // Update timestamp
    ghosts.updated_at = chrono::Utc::now().to_rfc3339();

    // Write back
    let output = serde_json::to_string_pretty(&ghosts)?;
    std::fs::write(&ghosts_path, output)?;

    println!("Updated reference-ghosts.json");

    Ok(())
}
